python/ql/src/experimental/Security/CWE-287-ConstantSecretKey/ConstantSecretKey.ql

```diff
@@ -0,0 +1,219 @@
+/**
+ * @name Initializing SECRET_KEY of Flask application with Constant value
+ * @description Initializing SECRET_KEY of Flask application with Constant value
+ * files can lead to Authentication bypass
+ * @kind path-problem
+ * @id py/ConstantSecretKey
+ * @problem.severity error
+ * @security-severity 8.5
+ * @precision high
+ * @tags security
+ *       experimental
+ *       external/cwe/cwe-287
+ */
+
+import python
+import experimental.semmle.python.Concepts
+import semmle.python.dataflow.new.DataFlow
+import semmle.python.ApiGraphs
+import semmle.python.dataflow.new.TaintTracking
+
+/**
+ * `flask.Flask()`
+ */
+API::Node flaskInstance() { result = API::moduleImport("flask").getMember("Flask").getASubclass*() }
+
+/**
+ * with using flask-session package, there is no jwt exists in cookies in user side
+ * ```python
+ *import os
+ *from flask import Flask, session
+ *app = Flask(__name__)
+ * ```
+ */
+module FlaskConstantSecretKeyConfig implements DataFlow::ConfigSig {
+  /**
+   * Sources are Constants that without any Tainting reach the Sinks.
+   * Also Sources can be the default value of getenv or similar methods
+   * in a case that no value is assigned to Desired SECRET_KEY environment variable
+   */
+  predicate isSource(DataFlow::Node source) {
+    (
+      source.asExpr().isConstant()
+      or
+      exists(API::Node cn |
+        cn =
+          [
+            API::moduleImport("configparser")
+                .getMember(["ConfigParser", "RawConfigParser"])
+                .getReturn(),
+            // legacy API https://docs.python.org/3/library/configparser.html#legacy-api-examples
+            API::moduleImport("configparser")
+                .getMember(["ConfigParser", "RawConfigParser"])
+                .getReturn()
+                .getMember("get")
+                .getReturn()
+          ] and
+        source = cn.asSource()
+      )
+      or
+      exists(API::CallNode cn |
+        cn =
+          [
+            API::moduleImport("os").getMember("getenv").getACall(),
+            API::moduleImport("os").getMember("environ").getMember("get").getACall()
+          ] and
+        (
+          // this can be ideal if we assume that best security practice is that
+          // we don't get SECRET_KEY from env and we always assign a secure generated random string to it
+          cn.getNumArgument() = 1
```

Comment:
As already outlined, I don't think this is best security practice. Do you have any resources to back up your claims? :blush:

Reply:
@amammad Sorry for the late response, I completely missed this thread. 

Storing secret keys in environment variables is a globally recognized good practice. Loading those variables from local files completely defeat the purpose of adding them to the OS environment at boot time. However, from a static analysis point of view, we dont have visibility on where the env vars are loaded from and it is ok to assume that loading secrets from env vars is a safe approach.

Reporting issues/vulnerabilities for each case where secrets are loaded from env vars will lead to too many FPs. Please adjust the query to not report those issues.

Reply:
The default values should be reported since they are potential hardcoded credentials